### PR TITLE
Fix #324

### LINKED ISSUE #324: Should this Set pickle-unpickle with val fail?
``` scala
scala> import scala.pickling.Defaults._, scala.pickling.json._
import scala.pickling.Defaults._
import scala.pickling.json._

scala> case class Foo(bar: Set[Int]) { val barSize = bar.size }
defined class Foo

scala> Foo(Set()).pickle
res27: pickling.json.pickleFormat.PickleType =
JSONPickle({
  "$type": "Foo",
  "bar": {
    "$type": "scala.collection.immutable.Set[scala.Int]",
    "elems": [

    ]
  }
})

scala> res27.unpickle[Foo]
java.lang.NullPointerException
  ... 49 elided

scala> lastException.printStackTrace
java.lang.NullPointerException
    at $line57.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$Foo.<init>(<console>:22)
    at $line59.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$line57$read$iw$iw$iw$iw$iw$iw$iw$iw$iw$iw$iw$iwFooUnpickler$macro$49$2$.unpickle(<console>:26)
    at scala.pickling.Unpickler$class.unpickleEntry(Pickler.scala:79)
    at $line59.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$line57$read$iw$iw$iw$iw$iw$iw$iw$iw$iw$iw$iw$iwFooUnpickler$macro$49$2$.unpickleEntry(<console>:26)
    at scala.pickling.functions$.unpickle(functions.scala:11)
    at scala.pickling.UnpickleOps.unpickle(Ops.scala:23)
    at $line59.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$.<init>(<console>:26)
    at $line59.$read$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$$iw$.<clinit>(<console>)
    at $line59.$eval$.$print$lzycompute(<console>:7)
    at $line59.$eval$.$print(<console>:6)
    at $line59.$eval.$print(<console>)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:483)
    at scala.tools.nsc.interpreter.IMain$ReadEvalPrint.call(IMain.scala:773)
    at scala.tools.nsc.interpreter.IMain$Request.loadAndRun(IMain.scala:1020)
    at scala.tools.nsc.interpreter.IMain$WrappedRequest$$anonfun$loadAndRunReq$1.apply(IMain.scala:627)
    at scala.tools.nsc.interpreter.IMain$WrappedRequest$$anonfun$loadAndRunReq$1.apply(IMain.scala:626)
    at scala.reflect.internal.util.ScalaClassLoader$class.asContext(ScalaClassLoader.scala:31)
    at scala.reflect.internal.util.AbstractFileClassLoader.asContext(AbstractFileClassLoader.scala:19)
    at scala.tools.nsc.interpreter.IMain$WrappedRequest.loadAndRunReq(IMain.scala:626)
    at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:558)
    at scala.tools.nsc.interpreter.IMain.interpret(IMain.scala:554)
    at scala.tools.nsc.interpreter.ILoop.reallyInterpret$1(ILoop.scala:804)
    at scala.tools.nsc.interpreter.ILoop.interpretStartingWith(ILoop.scala:849)
    at scala.tools.nsc.interpreter.ILoop.command(ILoop.scala:707)
    at scala.tools.nsc.interpreter.ILoop.processLine(ILoop.scala:405)
    at scala.tools.nsc.interpreter.ILoop.loop(ILoop.scala:431)
    at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply$mcZ$sp(ILoop.scala:920)
    at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply(ILoop.scala:906)
    at scala.tools.nsc.interpreter.ILoop$$anonfun$process$1.apply(ILoop.scala:906)
    at scala.reflect.internal.util.ScalaClassLoader$.savingContextLoader(ScalaClassLoader.scala:97)
    at scala.tools.nsc.interpreter.ILoop.process(ILoop.scala:906)
    at scala.tools.nsc.interpreter.ILoop.main(ILoop.scala:931)
    at xsbt.ConsoleInterface.run(ConsoleInterface.scala:69)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:483)
    at sbt.compiler.AnalyzingCompiler.call(AnalyzingCompiler.scala:102)
    at sbt.compiler.AnalyzingCompiler.console(AnalyzingCompiler.scala:77)
    at sbt.Console.sbt$Console$$console0$1(Console.scala:23)
    at sbt.Console$$anonfun$apply$2$$anonfun$apply$1.apply$mcV$sp(Console.scala:24)
    at sbt.Console$$anonfun$apply$2$$anonfun$apply$1.apply(Console.scala:24)
    at sbt.Console$$anonfun$apply$2$$anonfun$apply$1.apply(Console.scala:24)
    at sbt.Logger$$anon$4.apply(Logger.scala:90)
    at sbt.TrapExit$App.run(TrapExit.scala:244)
    at java.lang.Thread.run(Thread.java:745)
```

### PR DESCRIPTION
**Pull request by phaller**
Fix #324

### CODE REVIEW

File: core/src/main/scala/scala/pickling/Macros.scala [hunk 1]
```diff
@@ -497,7 +496,7 @@ trait UnpicklerMacros extends Macro with UnpickleMacros with FastTypeTagMacros {
       // in the worst case we can do the same as the interpreted runtime does - just go for allocateInstance
 
       // pending fields are fields that are restored after instantiation (e.g., through field assignments)
-      val pendingFields = if (!canCallCtor) cir.fields else cir.fields.filter(fir =>
+      var pendingFields = if (!canCallCtor) cir.fields else cir.fields.filter(fir =>
```
Comment (jsuereth) [inline core/src/main/scala/scala/pickling/Macros.scala:499]: I guess there's no way we can avoid having it be a `var`?

The only reason I ask is because this logic is already kind of hard to follow at first.  I'd love to take this chance to maybe clean up into helper methods that help delineate what we're doing...

**Comment by jsuereth** [comment]
The fix looks good.  Just a comment on trying to improve the flow of logic, if we can.

File: core/src/main/scala/scala/pickling/Macros.scala (hunk 1 above)
Comment (phaller) [inline core/src/main/scala/scala/pickling/Macros.scala:499]: @jsuereth I share your sentiment. :-) Cleaning up this logic would be nice. I submitted the PR in the present form mainly because I ran out of time, but still wanted to share my bug fix with people.

**Comment by eed3si9n** [comment]
As I'm planning to release 0.10.1, I'm going to merge as is but create a Github issue so @phaller can come back to it later.
